### PR TITLE
Netperf: add support for failover

### PR DESCRIPTION
**Pull request by quanwenli**
Add support for failover

Signed-off-by: Wenli Quan<wquan@redhat.com>

id: 2026566

### CODE REVIEW

**Comment by quanwenli** [comment]
JOB LOG    : /root/avocado/job-results/job-2021-11-25T21.46-1eb3986/job.log
 (1/1) 82599.failover_test.1q.acceptance.repeat1.Host_RHEL.m8.u5.product_av.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.8.5.0.x86_64.io-github-autotest-qemu.netperf.default.host_guest.q35: PASS (291.18 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2021-11-25T21.46-1eb3986/results.html
JOB TIME   : 292.79 s

**Comment by quanwenli** [comment]
For failover for windows support, there is an existed bug 1969724, so currently it's only test passed with "driver_name = netkvm"

JOB ID     : d6732866f69a29796783e906bd91a5e493dd5a85
JOB LOG    : /root/avocado/job-results/job-2021-12-01T22.05-d673286/job.log
 (1/1) 82599.failover_test.1q.acceptance.repeat1.Host_RHEL.m8.u5.product_av.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.netperf.default.host_guest.q35: PASS (419.52 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2021-12-01T22.05-d673286/results.html

**Comment by quanwenli** [comment]
@vivianQizhu could you help review it ?

**Comment by vivianQizhu** [comment]
@yanghangliu @yanglei-rh @PaulYuuu Would you help review this one?

**Comment by yanglei-rh** [comment]
The code currently changed does not involve netperf. In addition, I don’t know much about failover testing.

**Comment by yanghangliu** [comment]
@YanhuiMa Yanhui Could you please help review this patch ?

File: generic/tests/netperf.py [hunk 1]
```diff
@@ -118,10 +119,62 @@ def mtu_set(mtu):
...
+            session = vm.wait_for_serial_login(timeout=login_timeout)
+            ifname = utils_net.get_linux_ifname(session)
+            for i in ifname:
+                cmd = "ethtool -i %s |grep driver| awk -F': ' '{print $2}'" % i
+                driver = session.cmd_output(cmd).strip()
+                if driver == "net_failover":
+                    session.cmd_output("dhclient -r && dhclient %s" % i)
+                break
+        if params.get("os_type") == "windows" and params.get("install_vioprot_cmd"):
+            media_type = params.get("virtio_win_media_type")
```
Comment (YanhuiMa) [inline generic/tests/netperf.py:133]: It should be params["virtio_win_media_type"] here?

File: generic/tests/netperf.py [hunk 2]
```diff
@@ -118,10 +119,62 @@ def mtu_set(mtu):
...
+                                                      devcon_path, driver_name,
+                                                      device_name, device_hwid)
+                    for hwid in device_hwid.split():
+                        install_driver_cmd = "%s install %s %s" % (devcon_path,
+                                                                   inf_path,
+                                                                   hwid)
+                        status, output = session.cmd_status_output(
+                                         install_driver_cmd,
+                                         timeout=login_timeout)
+                        if status > 1:
```
Comment (YanhuiMa) [inline generic/tests/netperf.py:162]: Could we directly use "If status:"?

File: generic/tests/netperf.py [hunk 3]
```diff
@@ -118,10 +119,62 @@ def mtu_set(mtu):
         if params.get("reboot_after_config", "yes") == "yes":
             vm.reboot(method="system_reset", serial=True)
 
-    try:
-        vm.wait_for_serial_login(timeout=login_timeout, restart_network=True).close()
-    except virt_vm.VMIPAddressMissingError:
-        pass
+    if params.get("failover_pair_id_nic1"):
```
[the remark this answers is not in the document]
Comment (quanwenli) [reply generic/tests/netperf.py:122]: There is params failover_pair_id in virttest/qemu_vm.py, in my test , it's in my test loop.

File: generic/tests/cfg/netperf.cfg [hunk 4]
```diff
@@ -125,6 +130,13 @@
                         config_cmds = driver_verifier_query
                         driver_verifier_query = "verifier /querysettings"
                         reboot_after_config = yes
+                        install_vioprot_cmd = "netcfg -v -l %s -c p -i VIOPROT"
+                        cdroms += " virtio winutils"
+                        virtio_win_media_type = iso
+                        driver_name = netkvm VIOPROT
```
Comment (YanhuiMa) [inline generic/tests/cfg/netperf.cfg:136]: Do you need to add VF driver?

Comment (quanwenli) [reply generic/tests/cfg/netperf.cfg:136]: > Do you need to add VF driver?

Yes, we need add VF driver, but it needs download drivers from intel or other company who support SRIOV. currently, we can only manually installed vf drivers before tesing, so this part, I will research future about how to surpport it.

File: generic/tests/netperf.py (hunk 1 above)
Comment (quanwenli) [reply generic/tests/netperf.py]: updated.

File: generic/tests/netperf.py (hunk 2 above)
Comment (quanwenli) [reply generic/tests/netperf.py]: updated.

File: generic/tests/netperf.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (quanwenli) [reply generic/tests/netperf.py]: > We cannot make sure the nic name is always `nic1`, so here use `params.object_params(nic_name)` --> `params.get("failover_pair_id")` should be better.

I have udpated and test is passed, could you help review?

JOB LOG    : /root/avocado/job-results/job-2021-12-15T22.36-2a245a8/job.log
 (1/1) 82599.failover_test.1q.acceptance.repeat1.Host_RHEL.m9.u0.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.netperf.default.host_guest.q35: PASS (262.10 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2021-12-15T22.36-2a245a8/results.html
JOB TIME   : 263.15 s

File: generic/tests/netperf.py [hunk 5]
```diff
@@ -118,10 +119,67 @@ def mtu_set(mtu):
         if params.get("reboot_after_config", "yes") == "yes":
             vm.reboot(method="system_reset", serial=True)
 
-    try:
-        vm.wait_for_serial_login(timeout=login_timeout, restart_network=True).close()
-    except virt_vm.VMIPAddressMissingError:
-        pass
+    failover_exist = False
+    for i in range(1, len(params.get("nics", "").split()) + 1):
+        nic_params = params.object_params("nic%s" % i)
```
Comment (PaulYuuu) [inline generic/tests/netperf.py:124]:
```suggestion
    for i in params.objects("nics"):
        nic_params = params.object_params(i)
```

Comment (quanwenli) [reply generic/tests/netperf.py]: I have udpated, please help review. 

JOB LOG    : /root/avocado/job-results/job-2021-12-21T21.43-e06051f/job.log
 (1/1) 82599.failover_test.1q.acceptance.repeat1.Host_RHEL.m8.u6.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.netperf.default.host_guest.q35: PASS (389.75 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2021-12-21T21.43-e06051f/results.html

Comment (quanwenli) [reply generic/tests/netperf.py]: @PaulYuuu , any comments. thanks

File: generic/tests/netperf.py [hunk 6]
```diff
@@ -118,10 +119,67 @@ def mtu_set(mtu):
...
-    try:
-        vm.wait_for_serial_login(timeout=login_timeout, restart_network=True).close()
-    except virt_vm.VMIPAddressMissingError:
-        pass
+    failover_exist = False
+    for i in params.get("nics").split():
+        nic_params = params.object_params(i)
+        if nic_params.get("failover_pair_id"):
+            failover_exist = True
+    if failover_exist:
```
Comment (PaulYuuu) [inline generic/tests/netperf.py:127]: I am curious when there are multiple NICs, the value of `failover_exist` always be the latest one, what is this need?

Comment (quanwenli) [reply generic/tests/netperf.py:127]: > I am curious when there are multiple NICs, the value of `failover_exist` always be the latest one, what is this need?

The default, we test with muliple nics, for only get "nic_params.get("failover_pair_id")" the failover_exist will be ture, for other scenarios, like bridge test, the failover_exist = False.

Comment (PaulYuuu) [reply generic/tests/netperf.py:127]: So why get `failover_exist` repeatedly in a loop, the logic should be wrong.

Comment (quanwenli) [reply generic/tests/netperf.py]: > So why get `failover_exist` repeatedly in a loop, the logic should be wrong.

ok, add break in the loop. and test it, it's passed. 

JOB LOG    : /root/avocado/job-results/job-2022-01-07T02.24-dcf4bcd/job.log
 (1/1) 82599.failover_test.1q.acceptance.repeat1.Host_RHEL.m8.u6.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.netperf.default.host_guest.q35: PASS (315.54 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /root/avocado/job-results/job-2022-01-07T02.24-dcf4bcd/results.html

**Review by PaulYuuu — APPROVED** [review]

**Pull request merged**